{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20.3\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Cannot close a running event loop",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\telegram\\ext\\_application.py:921\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m    920\u001b[0m     updater_coroutine\u001b[39m.\u001b[39mclose()\n\u001b[1;32m--> 921\u001b[0m     \u001b[39mraise\u001b[39;00m exc\n\u001b[0;32m    922\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    923\u001b[0m     \u001b[39m# We arrive here either by catching the exceptions above or if the loop gets stopped\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\telegram\\ext\\_application.py:910\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m    909\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 910\u001b[0m     loop\u001b[39m.\u001b[39;49mrun_until_complete(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49minitialize())\n\u001b[0;32m    911\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpost_init:\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\base_events.py:622\u001b[0m, in \u001b[0;36mBaseEventLoop.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    621\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_closed()\n\u001b[1;32m--> 622\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_check_running()\n\u001b[0;32m    624\u001b[0m new_task \u001b[39m=\u001b[39m \u001b[39mnot\u001b[39;00m futures\u001b[39m.\u001b[39misfuture(future)\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\base_events.py:582\u001b[0m, in \u001b[0;36mBaseEventLoop._check_running\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    581\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mis_running():\n\u001b[1;32m--> 582\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\u001b[39m'\u001b[39m\u001b[39mThis event loop is already running\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m    583\u001b[0m \u001b[39mif\u001b[39;00m events\u001b[39m.\u001b[39m_get_running_loop() \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[1;31mRuntimeError\u001b[0m: This event loop is already running",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\telegram\\ext\\_application.py:932\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m    931\u001b[0m     loop\u001b[39m.\u001b[39mrun_until_complete(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpost_stop(\u001b[39mself\u001b[39m))\n\u001b[1;32m--> 932\u001b[0m loop\u001b[39m.\u001b[39;49mrun_until_complete(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mshutdown())\n\u001b[0;32m    933\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpost_shutdown:\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\base_events.py:622\u001b[0m, in \u001b[0;36mBaseEventLoop.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    621\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_closed()\n\u001b[1;32m--> 622\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_check_running()\n\u001b[0;32m    624\u001b[0m new_task \u001b[39m=\u001b[39m \u001b[39mnot\u001b[39;00m futures\u001b[39m.\u001b[39misfuture(future)\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\base_events.py:582\u001b[0m, in \u001b[0;36mBaseEventLoop._check_running\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    581\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mis_running():\n\u001b[1;32m--> 582\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\u001b[39m'\u001b[39m\u001b[39mThis event loop is already running\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m    583\u001b[0m \u001b[39mif\u001b[39;00m events\u001b[39m.\u001b[39m_get_running_loop() \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[1;31mRuntimeError\u001b[0m: This event loop is already running",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 165\u001b[0m\n\u001b[0;32m    161\u001b[0m     application\u001b[38;5;241m.\u001b[39mrun_polling()\n\u001b[0;32m    164\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m--> 165\u001b[0m     \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[1], line 161\u001b[0m, in \u001b[0;36mmain\u001b[1;34m()\u001b[0m\n\u001b[0;32m    158\u001b[0m application\u001b[38;5;241m.\u001b[39madd_handler(conv_handler)\n\u001b[0;32m    160\u001b[0m \u001b[38;5;66;03m# Run the bot until the user presses Ctrl-C\u001b[39;00m\n\u001b[1;32m--> 161\u001b[0m \u001b[43mapplication\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_polling\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\telegram\\ext\\_application.py:727\u001b[0m, in \u001b[0;36mApplication.run_polling\u001b[1;34m(self, poll_interval, timeout, bootstrap_retries, read_timeout, write_timeout, connect_timeout, pool_timeout, allowed_updates, drop_pending_updates, close_loop, stop_signals)\u001b[0m\n\u001b[0;32m    724\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39merror_callback\u001b[39m(exc: TelegramError) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    725\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcreate_task(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprocess_error(error\u001b[39m=\u001b[39mexc, update\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m))\n\u001b[1;32m--> 727\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__run(\n\u001b[0;32m    728\u001b[0m     updater_coroutine\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mupdater\u001b[39m.\u001b[39;49mstart_polling(\n\u001b[0;32m    729\u001b[0m         poll_interval\u001b[39m=\u001b[39;49mpoll_interval,\n\u001b[0;32m    730\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[0;32m    731\u001b[0m         bootstrap_retries\u001b[39m=\u001b[39;49mbootstrap_retries,\n\u001b[0;32m    732\u001b[0m         read_timeout\u001b[39m=\u001b[39;49mread_timeout,\n\u001b[0;32m    733\u001b[0m         write_timeout\u001b[39m=\u001b[39;49mwrite_timeout,\n\u001b[0;32m    734\u001b[0m         connect_timeout\u001b[39m=\u001b[39;49mconnect_timeout,\n\u001b[0;32m    735\u001b[0m         pool_timeout\u001b[39m=\u001b[39;49mpool_timeout,\n\u001b[0;32m    736\u001b[0m         allowed_updates\u001b[39m=\u001b[39;49mallowed_updates,\n\u001b[0;32m    737\u001b[0m         drop_pending_updates\u001b[39m=\u001b[39;49mdrop_pending_updates,\n\u001b[0;32m    738\u001b[0m         error_callback\u001b[39m=\u001b[39;49merror_callback,  \u001b[39m# if there is an error in fetching updates\u001b[39;49;00m\n\u001b[0;32m    739\u001b[0m     ),\n\u001b[0;32m    740\u001b[0m     close_loop\u001b[39m=\u001b[39;49mclose_loop,\n\u001b[0;32m    741\u001b[0m     stop_signals\u001b[39m=\u001b[39;49mstop_signals,\n\u001b[0;32m    742\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\telegram\\ext\\_application.py:937\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m    935\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    936\u001b[0m     \u001b[39mif\u001b[39;00m close_loop:\n\u001b[1;32m--> 937\u001b[0m         loop\u001b[39m.\u001b[39;49mclose()\n",
      "File \u001b[1;32mc:\\Users\\diku3\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\selector_events.py:84\u001b[0m, in \u001b[0;36mBaseSelectorEventLoop.close\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     82\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mclose\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m     83\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mis_running():\n\u001b[1;32m---> 84\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mCannot close a running event loop\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     85\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mis_closed():\n\u001b[0;32m     86\u001b[0m         \u001b[39mreturn\u001b[39;00m\n",
      "\u001b[1;31mRuntimeError\u001b[0m: Cannot close a running event loop"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "# pylint: disable=unused-argument, wrong-import-position\n",
    "# This program is dedicated to the public domain under the CC0 license.\n",
    "\n",
    "\"\"\"\n",
    "First, a few callback functions are defined. Then, those functions are passed to\n",
    "the Application and registered at their respective places.\n",
    "Then, the bot is started and runs until we press Ctrl-C on the command line.\n",
    "\n",
    "Usage:\n",
    "Example of a bot-user conversation using ConversationHandler.\n",
    "Send /start to initiate the conversation.\n",
    "Press Ctrl-C on the command line or send a signal to the process to stop the\n",
    "bot.\n",
    "\"\"\"\n",
    "\n",
    "import logging\n",
    "from typing import Dict\n",
    "\n",
    "from telegram import __version__ as TG_VER\n",
    "print(TG_VER)\n",
    "\n",
    "try:\n",
    "    from telegram import __version_info__\n",
    "except ImportError:\n",
    "    __version_info__ = (0, 0, 0, 0, 0)  # type: ignore[assignment]\n",
    "\n",
    "if __version_info__ < (20, 0, 0, \"alpha\", 1):\n",
    "    raise RuntimeError(\n",
    "        f\"This example is not compatible with your current PTB version {TG_VER}. To view the \"\n",
    "        f\"{TG_VER} version of this example, \"\n",
    "        f\"visit https://docs.python-telegram-bot.org/en/v{TG_VER}/examples.html\"\n",
    "    )\n",
    "from telegram import ReplyKeyboardMarkup, ReplyKeyboardRemove, Update\n",
    "from telegram.ext import (\n",
    "    Application,\n",
    "    CommandHandler,\n",
    "    ContextTypes,\n",
    "    ConversationHandler,\n",
    "    MessageHandler,\n",
    "    filters,\n",
    ")\n",
    "\n",
    "# Enable logging\n",
    "logging.basicConfig(\n",
    "    format=\"%(asctime)s - %(name)s - %(levelname)s - %(message)s\", level=logging.INFO\n",
    ")\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "CHOOSING, TYPING_REPLY, TYPING_CHOICE = range(3)\n",
    "\n",
    "reply_keyboard = [\n",
    "    [\"Age\", \"Favourite colour\"],\n",
    "    [\"Number of siblings\", \"Something else...\"],\n",
    "    [\"Done\"],\n",
    "]\n",
    "markup = ReplyKeyboardMarkup(reply_keyboard, one_time_keyboard=True)\n",
    "\n",
    "\n",
    "def facts_to_str(user_data: Dict[str, str]) -> str:\n",
    "    \"\"\"Helper function for formatting the gathered user info.\"\"\"\n",
    "    facts = [f\"{key} - {value}\" for key, value in user_data.items()]\n",
    "    return \"\\n\".join(facts).join([\"\\n\", \"\\n\"])\n",
    "\n",
    "\n",
    "async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:\n",
    "    \"\"\"Start the conversation and ask user for input.\"\"\"\n",
    "    await update.message.reply_text(\n",
    "        \"Hi! My name is Doctor Botter. I will hold a more complex conversation with you. \"\n",
    "        \"Why don't you tell me something about yourself?\",\n",
    "        reply_markup=markup,\n",
    "    )\n",
    "\n",
    "    return CHOOSING\n",
    "\n",
    "\n",
    "async def regular_choice(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:\n",
    "    \"\"\"Ask the user for info about the selected predefined choice.\"\"\"\n",
    "    text = update.message.text\n",
    "    context.user_data[\"choice\"] = text\n",
    "    await update.message.reply_text(f\"Your {text.lower()}? Yes, I would love to hear about that!\")\n",
    "\n",
    "    return TYPING_REPLY\n",
    "\n",
    "\n",
    "async def custom_choice(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:\n",
    "    \"\"\"Ask the user for a description of a custom category.\"\"\"\n",
    "    await update.message.reply_text(\n",
    "        'Alright, please send me the category first, for example \"Most impressive skill\"'\n",
    "    )\n",
    "\n",
    "    return TYPING_CHOICE\n",
    "\n",
    "\n",
    "async def received_information(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:\n",
    "    \"\"\"Store info provided by user and ask for the next category.\"\"\"\n",
    "    user_data = context.user_data\n",
    "    text = update.message.text\n",
    "    category = user_data[\"choice\"]\n",
    "    user_data[category] = text\n",
    "    del user_data[\"choice\"]\n",
    "\n",
    "    await update.message.reply_text(\n",
    "        \"Neat! Just so you know, this is what you already told me:\"\n",
    "        f\"{facts_to_str(user_data)}You can tell me more, or change your opinion\"\n",
    "        \" on something.\",\n",
    "        reply_markup=markup,\n",
    "    )\n",
    "\n",
    "    return CHOOSING\n",
    "\n",
    "\n",
    "async def done(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:\n",
    "    \"\"\"Display the gathered info and end the conversation.\"\"\"\n",
    "    user_data = context.user_data\n",
    "    if \"choice\" in user_data:\n",
    "        del user_data[\"choice\"]\n",
    "\n",
    "    await update.message.reply_text(\n",
    "        f\"I learned these facts about you: {facts_to_str(user_data)}Until next time!\",\n",
    "        reply_markup=ReplyKeyboardRemove(),\n",
    "    )\n",
    "\n",
    "    user_data.clear()\n",
    "    return ConversationHandler.END\n",
    "\n",
    "\n",
    "def main() -> None:\n",
    "    \"\"\"Run the bot.\"\"\"\n",
    "    # Create the Application and pass it your bot's token.\n",
    "    application = Application.builder().token(\"5885078983:AAF2_Pmh_FS3gDHdm98bcyj6NYgy0Jr97ac\").build()\n",
    "\n",
    "    # Add conversation handler with the states CHOOSING, TYPING_CHOICE and TYPING_REPLY\n",
    "    conv_handler = ConversationHandler(\n",
    "        entry_points=[CommandHandler(\"start\", start)],\n",
    "        states={\n",
    "            CHOOSING: [\n",
    "                MessageHandler(\n",
    "                    filters.Regex(\"^(Age|Favourite colour|Number of siblings)$\"), regular_choice\n",
    "                ),\n",
    "                MessageHandler(filters.Regex(\"^Something else...$\"), custom_choice),\n",
    "            ],\n",
    "            TYPING_CHOICE: [\n",
    "                MessageHandler(\n",
    "                    filters.TEXT & ~(filters.COMMAND | filters.Regex(\"^Done$\")), regular_choice\n",
    "                )\n",
    "            ],\n",
    "            TYPING_REPLY: [\n",
    "                MessageHandler(\n",
    "                    filters.TEXT & ~(filters.COMMAND | filters.Regex(\"^Done$\")),\n",
    "                    received_information,\n",
    "                )\n",
    "            ],\n",
    "        },\n",
    "        fallbacks=[MessageHandler(filters.Regex(\"^Done$\"), done)],\n",
    "    )\n",
    "\n",
    "    application.add_handler(conv_handler)\n",
    "\n",
    "    # Run the bot until the user presses Ctrl-C\n",
    "    application.run_polling()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "1faf759d177765deb55ee80ab4abb1efe18815b38afeaebc22989555973fc7cb"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
